crates/ruff/src/commands/completions/config.rs

```diff
@@ -0,0 +1,158 @@
+use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
+use itertools::Itertools;
+use std::str::FromStr;
+
+use ruff_workspace::{
+    options::Options,
+    options_base::{OptionField, OptionSet, OptionsMetadata, Visit},
+};
+
+#[derive(Default)]
+struct CollectOptionsVisitor {
+    values: Vec<(String, String)>,
+    parents: Vec<String>,
+}
+
+impl IntoIterator for CollectOptionsVisitor {
+    type Item = (String, String);
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.values.into_iter()
+    }
+}
+
+impl Visit for CollectOptionsVisitor {
+    fn record_set(&mut self, name: &str, group: OptionSet) {
+        let fully_qualified_name = self
+            .parents
+            .iter()
+            .map(String::as_str)
+            .chain(std::iter::once(name))
+            .collect::<Vec<_>>()
+            .join(".");
+
+        // Only add the set to completion list if it has it's own documentation.
+        self.values.push((
+            fully_qualified_name,
+            group
+                .documentation()
+                .unwrap_or("Print a list of available options")
+                .to_owned(),
+        ));
+
+        self.parents.push(name.to_owned());
+        group.record(self);
+        self.parents.pop();
+    }
+
+    fn record_field(&mut self, name: &str, field: OptionField) {
+        let fqn = self
+            .parents
+            .iter()
+            .map(String::as_str)
+            .chain(std::iter::once(name))
+            .collect::<Vec<_>>()
+            .join(".");
+
+        self.values.push((fqn, field.doc.to_owned()));
+    }
+}
+
+/// Opaque type used solely to enable tab completions
+/// for `ruff option [OPTION]` command.
+#[derive(Clone, Debug)]
+pub struct OptionString(String);
+
+impl From<String> for OptionString {
+    fn from(s: String) -> Self {
+        OptionString(s)
+    }
+}
+
+impl From<OptionString> for String {
+    fn from(o: OptionString) -> Self {
+        o.0
+    }
```

Comment:
nit: `o` as a single letter is confusing
```suggestion
    fn from(value: OptionString) -> Self {
        value.0
    }
```